File: internal/generators/credential_guard_dedup_test.go
package generators

import (
	"os"
	"path/filepath"
	"testing"
)

// ---------------------------------------------------------------------------
// ROADMAP-2026-08-06 Wave 3/ML-3A — InjectXHooks (project scope) must skip
// the credential-guard entry when the corresponding global-scope wiring
// (installed via `trackfw update harness --targets <tool>-credential-guard`,
// internal/generators/update.go) is already present, and must fail-open
// (fall back to the pre-ML-3A behavior of always adding the project-scope
// entry) if the global file is missing, unreadable, or unparseable.
// ---------------------------------------------------------------------------

// dedupFixtureHome creates an isolated $HOME and points HOME at it for the
// duration of the test.
func dedupFixtureHome(t *testing.T) string {
	t.Helper()
	home := t.TempDir()
	t.Setenv("HOME", home)
	return home
}

func dedupScriptPath(home string) string {
	return filepath.Join(home, ".trackfw", "scripts", "trackfw-credential-guard.sh")
}

func TestDedup_Claude_SkipsProjectEntryWhenGlobalInstalled(t *testing.T) {
	home := dedupFixtureHome(t)
	scriptPath := dedupScriptPath(home)
	helperWriteJSON(t, filepath.Join(home, ".claude", "settings.json"), map[string]interface{}{
		"hooks": map[string]interface{}{
			"PreToolUse": []interface{}{
				map[string]interface{}{
					"matcher": "Bash",
					"hooks":   []interface{}{map[string]interface{}{"type": "command", "command": scriptPath}},
				},
			},
		},
	})

	dir := t.TempDir()
	if err := InjectClaudeHooks(dir); err != nil {
		t.Fatalf("InjectClaudeHooks failed: %v", err)
	}

	data := helperReadJSON(t, filepath.Join(dir, ".claude", "settings.json"))
	if helperHasClaudeHook(data, "PreToolUse", "Bash", "scripts/trackfw-credential-guard.sh") {
		t.Error("project-scope credential-guard entry should have been skipped (global already installed)")
	}
	if helperHasClaudeHook(data, "PostToolUse", "Bash", "scripts/trackfw-credential-guard.sh") {
		t.Error("project-scope PostToolUse credential-guard entry should have been skipped")
	}
	if !helperHasClaudeHook(data, "PreToolUse", "AskUserQuestion", "$CLAUDE_PROJECT_DIR/scripts/trackfw-attention-signal.sh") {
		t.Error("attention-signal entry must still be added regardless of global credential-guard state")
	}
	if !helperHasClaudeHook(data, "PostToolUse", "AskUserQuestion", "$CLAUDE_PROJECT_DIR/scripts/trackfw-attention-cleanup.sh") {
		t.Error("attention-cleanup entry must still be added regardless of global credential-guard state")
	}
}

func TestDedup_Codex_SkipsProjectEntryWhenGlobalInstalled(t *testing.T) {
	home := dedupFixtureHome(t)
	scriptPath := dedupScriptPath(home)
	helperWriteJSON(t, filepath.Join(home, ".codex", "hooks.json"), map[string]interface{}{
		"hooks": map[string]interface{}{
			"PreToolUse": []interface{}{
				map[string]interface{}{
					"matcher": "Bash",
					"hooks":   []interface{}{map[string]interface{}{"type": "command", "command": scriptPath}},
				},
			},
		},
	})

	dir := t.TempDir()
	if err := InjectCodexHooks(dir); err != nil {
		t.Fatalf("InjectCodexHooks failed: %v", err)
	}

	data := helperReadJSON(t, filepath.Join(dir, ".codex", "hooks.json"))
	if helperHasClaudeHook(data, "PreToolUse", "Bash", codexGuardCmd) {
		t.Error("project-scope credential-guard entry should have been skipped (global already installed)")
	}
	if helperHasClaudeHook(data, "PostToolUse", "Bash", codexGuardCmd) {
		t.Error("project-scope PostToolUse credential-guard entry should have been skipped")
	}
	if !helperHasClaudeHook(data, "PermissionRequest", ".*", codexSignalCmd) {
		t.Error("attention-signal entry must still be added regardless of global credential-guard state")
	}
	if !helperHasClaudeHook(data, "PostToolUse", ".*", codexCleanupCmd) {
		t.Error("attention-cleanup entry must still be added regardless of global credential-guard state")
	}
}

func TestDedup_Gemini_SkipsProjectEntryWhenGlobalInstalled(t *testing.T) {
	home := dedupFixtureHome(t)
	scriptPath := dedupScriptPath(home)
	helperWriteJSON(t, filepath.Join(home, ".gemini", "settings.json"), map[string]interface{}{
		"hooks": map[string]interface{}{
			"BeforeTool": []interface{}{
				map[string]interface{}{
					"matcher": "run_shell_command",
					"hooks":   []interface{}{map[string]interface{}{"type": "command", "command": scriptPath}},
				},
			},
		},
	})

	dir := t.TempDir()
	if err := InjectGeminiHooks(dir); err != nil {
		t.Fatalf("InjectGeminiHooks failed: %v", err)
	}

	data := helperReadJSON(t, filepath.Join(dir, ".gemini", "settings.json"))
	if helperHasClaudeHook(data, "BeforeTool", "run_shell_command", geminiGuardCmd) {
		t.Error("project-scope credential-guard entry should have been skipped (global already installed)")
	}
	if helperHasClaudeHook(data, "AfterTool", "run_shell_command", geminiGuardCmd) {
		t.Error("project-scope AfterTool credential-guard entry should have been skipped")
	}
	if !helperHasClaudeHook(data, "Notification", "ToolPermission", geminiSignalCmd) {
		t.Error("attention-signal entry must still be added regardless of global credential-guard state")
	}
	if !helperHasClaudeHook(data, "AfterTool", "*", geminiCleanupCmd) {
		t.Error("attention-cleanup entry must still be added regardless of global credential-guard state")
	}
}

func TestDedup_Cursor_SkipsProjectEntryWhenGlobalInstalled(t *testing.T) {
	home := dedupFixtureHome(t)
	scriptPath := dedupScriptPath(home)
	helperWriteJSON(t, filepath.Join(home, ".cursor", "hooks.json"), map[string]interface{}{
		"version": 1,
		"hooks": map[string]interface{}{
			"beforeShellExecution": []interface{}{
				map[string]interface{}{"command": scriptPath},
			},
		},
	})

	dir := t.TempDir()
	if err := InjectCursorHooks(dir); err != nil {
		t.Fatalf("InjectCursorHooks failed: %v", err)
	}

	data := helperReadJSON(t, filepath.Join(dir, ".cursor", "hooks.json"))
	hooks, _ := data["hooks"].(map[string]interface{})
	before, _ := hooks["beforeShellExecution"].([]interface{})
	after, _ := hooks["afterShellExecution"].([]interface{})
	// The git-branch-guard entry (ROADMAP-2026-08-14 ML-3A) has no
	// global-scope dedup target yet, so it is always added regardless of
	// the global credential-guard dedup state exercised by this test —
	// exactly 1 beforeShellExecution entry (git-branch-guard), 0 for
	// credential-guard (skipped, global installed).
	if len(before) != 1 || before[0].(map[string]interface{})["command"] != "scripts/trackfw-git-branch-guard.sh" {
		t.Errorf("expected only the git-branch-guard beforeShellExecution entry (credential-guard skipped, global already installed), got %v", before)
	}
	if len(after) != 0 {
		t.Errorf("expected no project-scope afterShellExecution entries (global already installed), got %v", after)
	}
	pre, _ := hooks["preToolUse"].([]interface{})
	post, _ := hooks["postToolUse"].([]interface{})
	if len(pre) != 1 || pre[0].(map[string]interface{})["command"] != "scripts/trackfw-attention-signal.sh" {
		t.Errorf("attention-signal entry must still be added regardless of global credential-guard state, got %v", pre)
	}
	if len(post) != 1 || post[0].(map[string]interface{})["command"] != "scripts/trackfw-attention-cleanup.sh" {
		t.Errorf("attention-cleanup entry must still be added regardless of global credential-guard state, got %v", post)
	}
}

func TestDedup_Copilot_SkipsProjectEntryWhenGlobalInstalled(t *testing.T) {
	home := dedupFixtureHome(t)
	scriptPath := dedupScriptPath(home)
	helperWriteJSON(t, filepath.Join(home, ".copilot", "settings.json"), map[string]interface{}{
		"hooks": map[string]interface{}{
			"preToolUse": []interface{}{
				map[string]interface{}{"type": "command", "matcher": "bash", "bash": scriptPath, "cwd": ".", "timeoutSec": 10},
			},
		},
	})

	dir := t.TempDir()
	if err := InjectCopilotHooks(dir); err != nil {
		t.Fatalf("InjectCopilotHooks failed: %v", err)
	}

	data := helperReadJSON(t, filepath.Join(dir, ".github", "hooks", "trackfw-attention.json"))
	hooks, _ := data["hooks"].(map[string]interface{})
	pre, _ := hooks["preToolUse"].([]interface{})
	post, _ := hooks["postToolUse"].([]interface{})
	// preToolUse: attention-signal + git-branch-guard (ROADMAP-2026-08-14
	// ML-3A, no global-scope dedup target yet, always added) — credential-guard
	// is skipped (global installed), so exactly 2 entries.
	if len(pre) != 2 {
		t.Errorf("expected attention-signal + git-branch-guard entries in preToolUse (global credential-guard installed), got %v", pre)
	}
	if pre[0].(map[string]interface{})["bash"] != "scripts/trackfw-attention-signal.sh" {
		t.Errorf("expected preToolUse[0] to be the attention-signal entry, got %v", pre[0])
	}
	foundGitGuard := false
	for _, item := range pre {
		if item.(map[string]interface{})["bash"] == "scripts/trackfw-git-branch-guard.sh" {
			foundGitGuard = true
		}
	}
	if !foundGitGuard {
		t.Errorf("expected preToolUse to contain the git-branch-guard entry, got %v", pre)
	}
	if len(post) != 1 || post[0].(map[string]interface{})["bash"] != "scripts/trackfw-attention-cleanup.sh" {
		t.Errorf("expected only the attention-cleanup entry in postToolUse (global credential-guard installed), got %v", post)
	}
}

func TestDedup_Kiro_SkipsProjectEntryWhenGlobalInstalled(t *testing.T) {
	home := dedupFixtureHome(t)
	globalKiroPath := filepath.Join(home, ".kiro", "hooks", "trackfw-credential-guard.json")
	if err := os.MkdirAll(filepath.Dir(globalKiroPath), 0755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	if err := os.WriteFile(globalKiroPath, []byte(`{"version":"v1","hooks":[]}`), 0644); err != nil {
		t.Fatalf("write: %v", err)
	}

	dir := t.TempDir()
	if err := InjectKiroHooks(dir); err != nil {
		t.Fatalf("InjectKiroHooks failed: %v", err)
	}

	data := helperReadJSON(t, filepath.Join(dir, ".kiro", "hooks", "trackfw-attention.json"))
	hooks, _ := data["hooks"].([]interface{})
	if len(hooks) != 2 {
		t.Fatalf("expected only 2 hooks (signal, cleanup) when global credential-guard is installed, got %d: %v", len(hooks), hooks)
	}
	for _, h := range hooks {
		entry, _ := h.(map[string]interface{})
		name, _ := entry["name"].(string)
		if name == "trackfw-credential-guard-pre" || name == "trackfw-credential-guard-post" {
			t.Errorf("project-scope credential-guard hook %q should have been skipped (global already installed)", name)
		}
	}
}

// --- Fail-open: missing/corrupted global file must not disable the
// project-scope credential-guard entry. ---

func TestDedup_FailOpen_NoGlobalFile(t *testing.T) {
	dedupFixtureHome(t) // empty $HOME, no global files at all

	dir := t.TempDir()
	if err := InjectClaudeHooks(dir); err != nil {
		t.Fatalf("InjectClaudeHooks failed: %v", err)
	}

	data := helperReadJSON(t, filepath.Join(dir, ".claude", "settings.json"))
	if !helperHasClaudeHook(data, "PreToolUse", "Bash", "$CLAUDE_PROJECT_DIR/scripts/trackfw-credential-guard.sh") {
		t.Error("expected project-scope credential-guard entry to be added when no global file exists (fail-open)")
	}
}

func TestDedup_FailOpen_CorruptedGlobalFile(t *testing.T) {
	home := dedupFixtureHome(t)
	path := filepath.Join(home, ".claude", "settings.json")
	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	if err := os.WriteFile(path, []byte("{not valid json"), 0644); err != nil {
		t.Fatalf("write: %v", err)
	}

	dir := t.TempDir()
	if err := InjectClaudeHooks(dir); err != nil {
		t.Fatalf("InjectClaudeHooks failed: %v", err)
	}

	data := helperReadJSON(t, filepath.Join(dir, ".claude", "settings.json"))
	if !helperHasClaudeHook(data, "PreToolUse", "Bash", "$CLAUDE_PROJECT_DIR/scripts/trackfw-credential-guard.sh") {
		t.Error("expected project-scope credential-guard entry to be added when global file is corrupted (fail-open)")
	}
}

func TestDedup_FailOpen_UnreadableGlobalFile(t *testing.T) {
	home := dedupFixtureHome(t)
	path := filepath.Join(home, ".claude", "settings.json")
	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	if err := os.WriteFile(path, []byte(`{"hooks":{}}`), 0000); err != nil {
		t.Fatalf("write: %v", err)
	}
	t.Cleanup(func() { _ = os.Chmod(path, 0644) }) // allow t.TempDir cleanup

	dir := t.TempDir()
	if err := InjectClaudeHooks(dir); err != nil {
		t.Fatalf("InjectClaudeHooks failed: %v", err)
	}

	data := helperReadJSON(t, filepath.Join(dir, ".claude", "settings.json"))
	if !helperHasClaudeHook(data, "PreToolUse", "Bash", "$CLAUDE_PROJECT_DIR/scripts/trackfw-credential-guard.sh") {
		t.Error("expected project-scope credential-guard entry to be added when global file is unreadable (fail-open)")
	}
}

File: npm/tests/credential_guard_dedup.test.js
'use strict'

// ROADMAP-2026-08-06 Wave 3/ML-3A — injectXHooks (project scope) must skip the
// credential-guard entry when the corresponding global-scope wiring (installed
// via `trackfw update harness --targets <tool>-credential-guard`,
// npm/src/commands/update-harness.js) is already present, and must fail-open
// (fall back to the pre-ML-3A behavior of always adding the project-scope
// entry) if the global file is missing, unreadable, or unparseable.
//
// Mirrors internal/generators/credential_guard_dedup_test.go (Go).

const test = require('node:test')
const assert = require('node:assert/strict')
const fs = require('node:fs')
const os = require('node:os')
const path = require('node:path')
const {
  injectClaudeHooks,
  injectCodexHooks,
  injectGeminiHooks,
  injectCursorHooks,
  injectCopilotHooks,
  injectKiroHooks,
} = require('../src/generators/hooks')

function writeJSON(filePath, data) {
  fs.mkdirSync(path.dirname(filePath), { recursive: true })
  fs.writeFileSync(filePath, JSON.stringify(data, null, 2) + '\n', 'utf8')
}

function readJSON(filePath) {
  return JSON.parse(fs.readFileSync(filePath, 'utf8'))
}

function hasClaudeHook(data, event, matcher, command) {
  const arr = (data.hooks && data.hooks[event]) || []
  return arr.some(e => e && e.matcher === matcher && Array.isArray(e.hooks) && e.hooks.some(h => h && h.command === command))
}

let origHome
test.beforeEach(() => {
  origHome = process.env.HOME
})
test.afterEach(() => {
  process.env.HOME = origHome
})

function isolatedHome() {
  const home = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-dedup-home-'))
  process.env.HOME = home
  return home
}

function scriptPathFor(home) {
  return path.join(home, '.trackfw', 'scripts', 'trackfw-credential-guard.sh')
}

test('injectClaudeHooks skips project-scope credential-guard when global installed', () => {
  const home = isolatedHome()
  const scriptPath = scriptPathFor(home)
  writeJSON(path.join(home, '.claude', 'settings.json'), {
    hooks: { PreToolUse: [{ matcher: 'Bash', hooks: [{ type: 'command', command: scriptPath }] }] },
  })

  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-dedup-project-'))
  injectClaudeHooks(dir)

  const data = readJSON(path.join(dir, '.claude', 'settings.json'))
  assert.equal(hasClaudeHook(data, 'PreToolUse', 'Bash', 'scripts/trackfw-credential-guard.sh'), false)
  assert.equal(hasClaudeHook(data, 'PostToolUse', 'Bash', 'scripts/trackfw-credential-guard.sh'), false)
  assert.equal(hasClaudeHook(data, 'PreToolUse', 'AskUserQuestion', '$CLAUDE_PROJECT_DIR/scripts/trackfw-attention-signal.sh'), true)
  assert.equal(hasClaudeHook(data, 'PostToolUse', 'AskUserQuestion', '$CLAUDE_PROJECT_DIR/scripts/trackfw-attention-cleanup.sh'), true)
})

test('injectCodexHooks skips project-scope credential-guard when global installed', () => {
  const home = isolatedHome()
  const scriptPath = scriptPathFor(home)
  writeJSON(path.join(home, '.codex', 'hooks.json'), {
    hooks: { PreToolUse: [{ matcher: 'Bash', hooks: [{ type: 'command', command: scriptPath }] }] },
  })

  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-dedup-project-'))
  injectCodexHooks(dir)

  const data = readJSON(path.join(dir, '.codex', 'hooks.json'))
  // ROADMAP-2026-08-11 ML-3A: Codex commands now wrap $(git rev-parse --show-toplevel) in literal
  // quotes (ADR-2026-08-11) -- see CODEX_*_CMD in generators.test.js for the shared literal.
  const codexGuardCmd = '"$(git rev-parse --show-toplevel)/scripts/trackfw-credential-guard.sh"'
  const codexSignalCmd = '"$(git rev-parse --show-toplevel)/scripts/trackfw-attention-signal.sh"'
  const codexCleanupCmd = '"$(git rev-parse --show-toplevel)/scripts/trackfw-attention-cleanup.sh"'
  assert.equal(hasClaudeHook(data, 'PreToolUse', 'Bash', codexGuardCmd), false)
  assert.equal(hasClaudeHook(data, 'PostToolUse', 'Bash', codexGuardCmd), false)
  assert.equal(hasClaudeHook(data, 'PermissionRequest', '.*', codexSignalCmd), true)
  assert.equal(hasClaudeHook(data, 'PostToolUse', '.*', codexCleanupCmd), true)
})

test('injectGeminiHooks skips project-scope credential-guard when global installed', () => {
  const home = isolatedHome()
  const scriptPath = scriptPathFor(home)
  writeJSON(path.join(home, '.gemini', 'settings.json'), {
    hooks: { BeforeTool: [{ matcher: 'run_shell_command', hooks: [{ type: 'command', command: scriptPath }] }] },
  })

  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-dedup-project-'))
  injectGeminiHooks(dir)

  const data = readJSON(path.join(dir, '.gemini', 'settings.json'))
  // ROADMAP-2026-08-11 ML-4A: Gemini commands now use $GEMINI_PROJECT_DIR (ADR-2026-08-11) --
  // see GEMINI_*_CMD in generators.test.js for the shared literal.
  const geminiGuardCmd = '$GEMINI_PROJECT_DIR/scripts/trackfw-credential-guard.sh'
  const geminiSignalCmd = '$GEMINI_PROJECT_DIR/scripts/trackfw-attention-signal.sh'
  const geminiCleanupCmd = '$GEMINI_PROJECT_DIR/scripts/trackfw-attention-cleanup.sh'
  assert.equal(hasClaudeHook(data, 'BeforeTool', 'run_shell_command', geminiGuardCmd), false)
  assert.equal(hasClaudeHook(data, 'AfterTool', 'run_shell_command', geminiGuardCmd), false)
  assert.equal(hasClaudeHook(data, 'Notification', 'ToolPermission', geminiSignalCmd), true)
  assert.equal(hasClaudeHook(data, 'AfterTool', '*', geminiCleanupCmd), true)
})

test('injectCursorHooks skips project-scope credential-guard when global installed', () => {
  const home = isolatedHome()
  const scriptPath = scriptPathFor(home)
  writeJSON(path.join(home, '.cursor', 'hooks.json'), {
    version: 1,
    hooks: { beforeShellExecution: [{ command: scriptPath }] },
  })

  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-dedup-project-'))
  injectCursorHooks(dir)

  const data = readJSON(path.join(dir, '.cursor', 'hooks.json'))
  // ROADMAP-2026-08-17 Wave 2/ML-2B: git-branch-guard now has its own global dedup
  // (globalGitBranchGuardInstalledCursor), but this fixture only wired the GLOBAL
  // credential-guard entry, not git-branch-guard's -- so the git-branch-guard dedup
  // check finds no matching command and fails open, keeping its project-scope entry.
  // See git_branch_guard_dedup.test.js for the case where the git-branch-guard global IS
  // installed.
  assert.equal((data.hooks.beforeShellExecution || []).length, 1)
  assert.equal(data.hooks.beforeShellExecution[0].command, 'scripts/trackfw-git-branch-guard.sh')
  assert.equal((data.hooks.afterShellExecution || []).length, 0)
  assert.equal(data.hooks.preToolUse.length, 1)
  assert.equal(data.hooks.preToolUse[0].command, 'scripts/trackfw-attention-signal.sh')
  assert.equal(data.hooks.postToolUse.length, 1)
  assert.equal(data.hooks.postToolUse[0].command, 'scripts/trackfw-attention-cleanup.sh')
})

test('injectCopilotHooks skips project-scope credential-guard when global installed', () => {
  const home = isolatedHome()
  const scriptPath = scriptPathFor(home)
  writeJSON(path.join(home, '.copilot', 'settings.json'), {
    hooks: { preToolUse: [{ type: 'command', matcher: 'bash', bash: scriptPath, cwd: '.', timeoutSec: 10 }] },
  })

  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-dedup-project-'))
  injectCopilotHooks(dir)

  const data = readJSON(path.join(dir, '.github', 'hooks', 'trackfw-attention.json'))
  // ROADMAP-2026-08-17 Wave 2/ML-2B: git-branch-guard now has its own global dedup
  // (globalGitBranchGuardInstalledCopilot), but this fixture only wired the GLOBAL
  // credential-guard entry, not git-branch-guard's -- so the git-branch-guard dedup
  // check finds no matching command and fails open, keeping its project-scope entry
  // alongside the always-on attention-signal entry.
  assert.equal(data.hooks.preToolUse.length, 2)
  assert.equal(data.hooks.preToolUse[0].bash, 'scripts/trackfw-attention-signal.sh')
  assert.equal(data.hooks.preToolUse[1].bash, 'scripts/trackfw-git-branch-guard.sh')
  assert.equal(data.hooks.postToolUse.length, 1)
  assert.equal(data.hooks.postToolUse[0].bash, 'scripts/trackfw-attention-cleanup.sh')
})

test('injectKiroHooks skips project-scope credential-guard when global installed', () => {
  const home = isolatedHome()
  const globalKiroPath = path.join(home, '.kiro', 'hooks', 'trackfw-credential-guard.json')
  writeJSON(globalKiroPath, { version: 'v1', hooks: [] })

  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-dedup-project-'))
  injectKiroHooks(dir)

  const data = readJSON(path.join(dir, '.kiro', 'hooks', 'trackfw-attention.json'))
  assert.equal(data.hooks.length, 2)
  for (const h of data.hooks) {
    assert.notEqual(h.name, 'trackfw-credential-guard-pre')
    assert.notEqual(h.name, 'trackfw-credential-guard-post')
  }
})

// --- Fail-open ---

test('injectClaudeHooks fail-open: no global file -> project entry still added', () => {
  isolatedHome() // empty $HOME, no global files at all

  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-dedup-failopen-'))
  injectClaudeHooks(dir)

  const data = readJSON(path.join(dir, '.claude', 'settings.json'))
  assert.equal(hasClaudeHook(data, 'PreToolUse', 'Bash', '$CLAUDE_PROJECT_DIR/scripts/trackfw-credential-guard.sh'), true)
})

test('injectClaudeHooks fail-open: corrupted global file -> project entry still added', () => {
  const home = isolatedHome()
  const settingsPath = path.join(home, '.claude', 'settings.json')
  fs.mkdirSync(path.dirname(settingsPath), { recursive: true })
  fs.writeFileSync(settingsPath, '{not valid json', 'utf8')

  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-dedup-failopen-'))
  injectClaudeHooks(dir)

  const data = readJSON(path.join(dir, '.claude', 'settings.json'))
  assert.equal(hasClaudeHook(data, 'PreToolUse', 'Bash', '$CLAUDE_PROJECT_DIR/scripts/trackfw-credential-guard.sh'), true)
})
